src/ray/gcs/pubsub/gcs_pub_sub.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gcs_pub_sub.h"
+#include "ray/gcs/redis_context.h"
+
+namespace ray {
+namespace gcs {
+
+template <typename ID, typename Data>
+Status GcsPubSub<ID, Data>::Publish(const std::string &channel, const ID &id,
+                                    const Data &data, const StatusCallback &done) {
+  rpc::GcsMessage message;
+  message.set_id(id.Binary());
+  std::string data_str;
+  data.SerializeToString(&data_str);
+  message.set_data(data_str);
+
+  auto on_done = [done](std::shared_ptr<CallbackReply> reply) {
+    if (done) {
+      done(Status::OK());
+    }
+  };
+
+  return redis_client_->GetPrimaryContext()->PublishAsync(
+      GenChannelPattern(channel, id), message.SerializeAsString(), on_done);
+}
+
+template <typename ID, typename Data>
+Status GcsPubSub<ID, Data>::Subscribe(const std::string &channel, const ID &id,
+                                      const Callback &subscribe,
+                                      const StatusCallback &done) {
+  return Subscribe(channel, boost::optional<ID>(id), subscribe, done);
+}
+
+template <typename ID, typename Data>
+Status GcsPubSub<ID, Data>::SubscribeAll(const std::string &channel,
+                                         const Callback &subscribe,
+                                         const StatusCallback &done) {
+  return Subscribe(channel, boost::none, subscribe, done);
+}
+
+template <typename ID, typename Data>
+Status GcsPubSub<ID, Data>::Unsubscribe(const std::string &channel, const ID &id,
+                                        const StatusCallback &done) {
+  if (done) {
+    unsubscribe_callbacks_[GenChannelPattern(channel, id)] = done;
```

Comment:
this is a bit weird to save the done callback and call it in `Subscribe`'s callback. 
Why not just call `done` directly here?

Reply:
fixed